src/task.c

```diff
@@ -1115,6 +1115,46 @@ JL_DLLEXPORT jl_task_t *jl_get_current_task(void)
     return pgcstack == NULL ? NULL : container_of(pgcstack, jl_task_t, gcstack);
 }
 
+extern int gc_first_tid;
+
+// Select a task at random to profile. Racy: `live_tasks` can change at any time.
+jl_task_t *jl_get_random_task(void) JL_NOTSAFEPOINT
+{
+    arraylist_t tasks;
+    arraylist_new(&tasks, 0);
+    size_t nthreads = jl_atomic_load_acquire(&jl_n_threads);
+    jl_ptls_t *allstates = jl_atomic_load_relaxed(&jl_all_tls_states);
+    for (size_t i = 0; i < nthreads; i++) {
+        // skip GC threads...
+        if (gc_first_tid <= i && i < gc_first_tid + jl_n_gcthreads) {
+            continue;
+        }
+        jl_ptls_t ptls2 = allstates[i];
+        if (ptls2 == NULL) {
+            continue;
+        }
+        jl_task_t *t = ptls2->root_task;
+        if (t->stkbuf != NULL) {
+            arraylist_push(&tasks, t);
+        }
+        small_arraylist_t *live_tasks = &ptls2->gc_tls.heap.live_tasks;
+        size_t n = mtarraylist_length(live_tasks);
+        for (size_t i = 0; i < n; i++) {
+            jl_task_t *t = (jl_task_t*)mtarraylist_get(live_tasks, i);
+            if (t->stkbuf != NULL) {
+                arraylist_push(&tasks, t);
+            }
+        }
+    }
+    size_t n = tasks.len;
+    if (n == 0) {
+        return NULL;
+    }
+    jl_task_t *t = (jl_task_t*)tasks.items[jl_rand() % n];
+    arraylist_free(&tasks);
+    return t;
```

Comment:
Did you benchmark this? (e.g. compared to the earlier version where we don't create an array for all tasks but instead loop over the threads once to get the total tasks and then again to retrieve the random task)

Reply:
Doesn't seem particularly concerning from a performance point of view.

## Julia Patch

```patch
diff --git a/src/signals-unix.c b/src/signals-unix.c
index db425a082c..32b935c8a9 100644
--- a/src/signals-unix.c
+++ b/src/signals-unix.c
@@ -44,6 +44,9 @@
 // Should also be enough for parallel GC when we have it =)
 const size_t sig_stack_size = (8 * 1024 * 1024);
 
+JL_DLLEXPORT size_t jl_cycles_in_get_random_task = 0;
+JL_DLLEXPORT size_t jl_cycles_in_profile_task = 0;
+
 #include "julia_assert.h"
 
 // helper function for returning the unw_context_t inside a ucontext_t
@@ -959,7 +962,10 @@ static void *signal_listener(void *arg)
         else if (profile) {
             if (profile_all_tasks) {
                 // Don't take the stackwalk lock here since it's already taken in `jl_rec_backtrace`
+                uint64_t c0 = cycleclock();
                 jl_profile_task_unix(nthreads);
+                uint64_t c1 = cycleclock();
+                jl_cycles_in_profile_task += c1 - c0;
             }
             else {
                 int lockret = jl_lock_stackwalk();
diff --git a/src/task.c b/src/task.c
index c9eb5f14ef..2208938323 100644
--- a/src/task.c
+++ b/src/task.c
@@ -1116,10 +1116,12 @@ JL_DLLEXPORT jl_task_t *jl_get_current_task(void)
 }
 
 extern int gc_first_tid;
+extern JL_DLLEXPORT size_t jl_cycles_in_get_random_task;
 
 // Select a task at random to profile. Racy: `live_tasks` can change at any time.
 jl_task_t *jl_get_random_task(void) JL_NOTSAFEPOINT
 {
+    uint64_t c0 = cycleclock();
     arraylist_t tasks;
     arraylist_new(&tasks, 0);
     size_t nthreads = jl_atomic_load_acquire(&jl_n_threads);
@@ -1152,6 +1154,8 @@ jl_task_t *jl_get_random_task(void) JL_NOTSAFEPOINT
     }
     jl_task_t *t = (jl_task_t*)tasks.items[jl_rand() % n];
     arraylist_free(&tasks);
+    uint64_t c1 = cycleclock();
+    jl_cycles_in_get_random_task += c1 - c0;
     return t;
 }
```

## MWE

```Julia
using Base.Threads
using Printf

function main()
    s = Threads.Atomic{Float64}(0)
    Threads.@sync begin
        Threads.@spawn begin
            for _ in 1:1_000
                Threads.atomic_add!(s, maximum(randn(1_000_000)))
            end
        end
    end
    return s
end

using Profile
@profile_all main()

# Load the C variable cycles_in_get_random_task
cycles_in_get_random_task = cglobal(:jl_cycles_in_get_random_task, UInt64)
cycles_in_jl_profile_task = cglobal(:jl_cycles_in_profile_task, UInt64)

Printf.@printf("ratio: %f\n", unsafe_load(cycles_in_get_random_task) / unsafe_load(cycles_in_jl_profile_task))
```

Gives:

```
ratio: 0.015434
```

I.e. 1.5% of time in `jl_profile_task_unix ` is spent in `jl_get_random_task`.

Reply:
Collecting backtraces tends to be quite expensive and probably dominates the cost here...